verus/lcf/src/kernel.rs

```diff
@@ -0,0 +1,1108 @@
+    pub open spec fn spec_valid(self) -> bool {
+        if self.spec_symbolic() || !(self.spec_concrete()) {
+            false
+        } else {
+            match self {
+                SpecProp::Eq(x, y) => match (x, y) {
+                    (SpecTerm::Const(x), SpecTerm::Const(y)) => x == y,
+                    (SpecTerm::Var(x), SpecTerm::Var(y)) => false,
+                    (SpecTerm::Const(_), SpecTerm::Var(_))
+                    | (SpecTerm::Var(_), SpecTerm::Const(_)) => false,
+                },
+                SpecProp::App(s, v) => false,
+            }
+        }
+    }
+
+    //reasons about substitution of SpecProp variants to SpecTerms
+    pub open spec fn spec_subst(self, s: SpecSubst) -> (res: SpecProp)
+        recommends
+            self.spec_complete_subst(s),
+    {
+        match self {
+            SpecProp::App(h, args) => {
+                let new_sequence = args.map_values(|p: SpecTerm| p.spec_subst(s));
+                SpecProp::App(h, new_sequence)
+            },
+            SpecProp::Eq(x, y) => SpecProp::Eq(x.spec_subst(s), y.spec_subst(s)),
+        }
+    }
+}
+
+impl Prop {
+    //proves equality for Prop types. uses term_eq and terms_eq
+    pub fn prop_eq(&self, other: &Self) -> (res: bool)
+        ensures
+            res <==> self.deep_view() == other.deep_view(),
+    {
+        match (self, other) {
+            (Prop::App(a, b), Prop::App(c, d)) => a.eq(c) && terms_eq(b, d),
+            (Prop::Eq(h, i), Prop::Eq(j, k)) => Term::term_eq(h, j) && Term::term_eq(i, k),
+            _ => false,
+        }
+    }
+
+    //checks whether Prop variants are concrete and elements are equal to each other
+    pub fn valid(self) -> (res: bool)
+        requires
+            !self.deep_view().spec_symbolic(),
+            self.deep_view().spec_concrete(),
+        ensures
+            res <==> self.deep_view().spec_valid(),
+    {
+        match self {
+            Prop::Eq(x, y) => match (x, y) {
+                //const_eq was needed here because the verifier cannot evaluate x == y on its own
+                (Term::Const(x), Term::Const(y)) => Const::const_eq(&x, &y),
+                (Term::Var(x), Term::Var(y)) => false,
+                (Term::Const(_), Term::Var(_)) | (Term::Var(_), Term::Const(_)) => false,
+            },
+            Prop::App(s, v) => false,
+        }
+    }
+
+    //checks whether Prop is an App variant
+    pub fn symbolic(self) -> (res: bool)
+        ensures
+            res <==> self.deep_view().spec_symbolic(),
+    {
+        match self {
+            Prop::App(_, _) => true,
+            Prop::Eq(_, _) => false,
+        }
+    }
+
+    //checks whether Prop variants are base types
+    pub fn concrete(self) -> (res: bool)
+        ensures
+            res <==> self.deep_view().spec_concrete(),
+    {
+        match self {
+            Prop::App(head, args) => {
+                assert(match self.deep_view() {
```

Comment:
This assertion also appears in `Prop::complete_subst` below, and possibly elsewhere. Eventually it would be good to factor it out into a lemma. Not important to do it now, but you could maybe leave a comment to note this.